### ISSUE TITLE
Data<T>() returns span that must be indexed using strides.

### ISSUE THREAD

**Issue by NiklasGustafsson** [dotnet/TorchSharp#399]
As reported by @shonk-msft:

This shows the issue:

```C#
using System;
using System.Text;
using TorchSharp;

class Program
{
    static void Main(string[] args)
    {
        // Create a contiguous 3x4 matrix and fill with auto-inc values starting at zero.
        // 0 1 2 3
        // 4 5 6 7
        // 8 9 10 11
        using var contig = torch.zeros(3, 4, torch.ScalarType.Int32);
        var data = contig.Data<int>();
        for (int i = 0; i < data.Length; i++)
            data[i] = i;
        Console.WriteLine(DumpData(contig));

        // Create a 3x2 slice of this, which should contain:
        // 1 2
        // 5 6
        // 9 10
        // Dumping values using Data<int>() shows incorrect values.
        using var sub = contig.slice(1, 1, 3, 1);
        Console.WriteLine(DumpData(sub));

        // Dumping sub.contiguous() shows the correct values.
        using var subContig = sub.contiguous();
        Console.WriteLine(DumpData(subContig));
    }

    static string DumpData(torch.Tensor ten)
    {
        var sb = new StringBuilder();
        sb.AppendFormat("Rank: {0}", ten.ndim).AppendLine();

        sb.Append("Shape: (");
        string pre = "";
        foreach (var dim in ten.shape)
        {
            sb.Append(pre).Append(dim);
            pre = ", ";
        }
         sb.AppendLine(")");

         sb.Append("Stride: (");
         pre = "";
         foreach (var dim in ten.stride())
         {
             sb.Append(pre).Append(dim);
             pre = ", ";
         }
         sb.AppendLine(")");

         var data = ten.Data<int>();
         sb.Append("Length: ").Append(data.Length).AppendLine();
         pre = "";
         for (int i = 0; i < data.Length; i++)
         {
             sb.Append(pre).Append(data[i]);
             pre = ", ";
         }
         sb.AppendLine();

         return sb.ToString();
    }
}
```
Running this produces:

```
Rank: 2
Shape: (3, 4)
Stride: (4, 1)
Length: 12
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11

Rank: 2
Shape: (3, 2)
Stride: (4, 1)
Length: 6
1, 2, 3, 4, 5, 6

Rank: 2
Shape: (3, 2)
Stride: (2, 1)
Length: 6
1, 2, 5, 6, 9, 10
```

The second dump's values should match the third's values, but they don't match because the Data<int>() provides a view on the contiguous data, not the data that actually belongs to the tensor.

Adding these few lines demonstrates the same issue using t rather than slice:

```C#
// Dumping the transpose.
using var trans = contig.t();
Console.WriteLine(DumpData(trans));
using var transContig = trans.contiguous();
Console.WriteLine(DumpData(transContig));
```

produces:

```
Rank: 2
Shape: (4, 3)
Stride: (1, 4)
Length: 12
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11

Rank: 2
Shape: (4, 3)
Stride: (3, 1)
Length: 12
0, 4, 8, 1, 5, 9, 2, 6, 10, 3, 7, 11
```

Again, the data values for the two should be the same (if one assumes that Data<int>() produced a view onto just this tensor's data).

Note that the issue is also exhibited by the ReadCpuXxx methods, since they just use Data<T>() under the covers.

Some possible safe solutions come to mind:

Make Data<T>() illegal on non-contiguous tensors. This is less than ideal, but safe.

In addition to this, fix ReadCpuXxx functions to properly compute the real index using the stride/shape information.

And introduce a ReadCpuValue<T>(long i) that allows one to stay in generic land.

Make Data<T>() (or a new function) return a logical-span that knows and uses the strides to map from index to correct location in the raw buffer. The main advantage of this over using ReadCpuValue<T>(long i) would be if the returned type implements IEnumerable<T> - the computation from logical index to physical index then becomes incremental rather than starting from scratch on each item.

I haven't used TorchSharp enough to find a way to get negative (or zero) stride values. Numpy allows that. With such strides, the span produced by Data<T>() can point off into illegal memory, so the issue becomes much worse.
